File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/migration/AccountDowngrader.kt
package com.bissbilanz.migration

import com.bissbilanz.api.BissbilanzApi
import com.bissbilanz.api.generated.model.EntryRangeItem
import com.bissbilanz.auth.AuthManager
import com.bissbilanz.mode.AppMode
import com.bissbilanz.mode.AppModeManager
import com.bissbilanz.model.Entry
import com.bissbilanz.sync.SyncQueue
import com.bissbilanz.userdata.UserDataDatabase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import kotlinx.datetime.Clock
import kotlinx.datetime.DatePeriod
import kotlinx.datetime.LocalDate
import kotlinx.datetime.TimeZone
import kotlinx.datetime.minus
import kotlinx.datetime.todayIn
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Turns a synced account into a local-only install: downloads the complete
 * account history into the local database, then deletes the server account and
 * flips the app to [AppMode.LOCAL] — the user keeps all their data on-device.
 *
 * The inverse of [LocalDataMigrator] (which uploads local data on sign-in).
 * Because the download writes canonical server rows (real UUIDs, no temp_ ids),
 * a later sign-in simply migrates them back up through the normal migration
 * path.
 */
class AccountDowngrader(
    private val api: BissbilanzApi,
    private val db: UserDataDatabase,
    private val syncQueue: SyncQueue,
    private val authManager: AuthManager,
    private val appModeManager: AppModeManager,
    private val json: Json,
    private val photoLocalizer: PhotoLocalizer? = null,
) {
    /** Localizes a server-hosted photo so it survives account deletion. */
    fun interface PhotoLocalizer {
        /**
         * Downloads the server-relative image (an `/uploads/…` URL) and returns
         * a replacement URL (e.g. `file://…`), or null to keep the original.
         */
        suspend fun localize(imageUrl: String): String?
    }

    /** Un-uploaded writes that must be flushed before the download starts. */
    suspend fun pendingOps(): Long = syncQueue.pendingCount()

    /**
     * Downloads every collection into the local database. Idempotent — every
     * write is an upsert of canonical server state, so a failed run can simply
     * be retried. Must complete without throwing before [finalize] is called.
     */
    suspend fun downloadAll(onProgress: (DownloadStep) -> Unit = {}) {
        withContext(Dispatchers.IO) {
            val account = api.getAccount()
            val today = Clock.System.todayIn(TimeZone.currentSystemDefault()).toString()
            val createdOn = account.user.createdAt?.take(10)
            // The account's own dated rows bound the download, NOT its creation
            // date: entry/sleep/weight/day dates are client-chosen, so imported
            // or backfilled days legitimately predate the account, and a device
            // running ahead of UTC can log a day past the server's "today".
            val floor = listOfNotNull(account.dataRange.earliest, createdOn).minOrNull() ?: FALLBACK_FLOOR
            val ceiling = maxOf(account.dataRange.latest ?: today, today)

            onProgress(DownloadStep.FOODS)
            downloadFoods()
            onProgress(DownloadStep.RECIPES)
            downloadRecipes()
            onProgress(DownloadStep.SUPPLEMENTS)
            downloadSupplements()
            onProgress(DownloadStep.ENTRIES)
            downloadDatedCollections(floor, ceiling)
            onProgress(DownloadStep.MEASUREMENTS)
            downloadWeight()
            onProgress(DownloadStep.SETTINGS)
            downloadGoals()
            downloadPreferences()
        }
    }

    /**
     * Deletes the server account and flips the app to Local mode. The local
     * database is left untouched — it is now the primary store.
     *
     * The mode flip happens BEFORE the irreversible delete: if the delete
     * succeeds but the process dies (or a later step throws) the app must never
     * be left in [AppMode.SYNCED] talking to an account that no longer exists —
     * from there the user's obvious next move, signing out, wipes the very data
     * the download just rescued. A failed delete restores the previous mode, so
     * the only state that survives an error is the one the user started in.
     */
    suspend fun finalize() {
        val previousMode = appModeManager.mode.value
        appModeManager.setMode(AppMode.LOCAL)
        try {
            api.deleteAccount()
        } catch (e: Throwable) {
            if (previousMode == null) appModeManager.clear() else appModeManager.setMode(previousMode)
            throw e
        }
        // The account is gone; the local store is now primary. Nothing below may
        // fail the downgrade — a stale queue row or session token is recoverable,
        // reporting failure here is not.
        runCatching { syncQueue.clear() }
        runCatching { authManager.logout() }
    }

    enum class DownloadStep { FOODS, RECIPES, SUPPLEMENTS, ENTRIES, MEASUREMENTS, SETTINGS }

    private suspend fun localizedImageUrl(imageUrl: String?): String? {
        if (imageUrl == null || !imageUrl.startsWith("/uploads/")) return imageUrl
        val localizer = photoLocalizer ?: return imageUrl
        return localizer.localize(imageUrl) ?: imageUrl
    }

    private suspend fun downloadFoods() {
        val queries = db.userDataDatabaseQueries
        var offset = 0
        while (true) {
            val page = api.getFoods(limit = PAGE_SIZE, offset = offset)
            val localized = page.map { it.copy(imageUrl = localizedImageUrl(it.imageUrl)) }
            queries.transaction {
                for (food in localized) {
                    queries.insertFood(
                        id = food.id,
                        name = food.name,
                        brand = food.brand,
                        calories = food.calories,
                        protein = food.protein,
                        carbs = food.carbs,
                        fat = food.fat,
                        fiber = food.fiber,
                        isFavorite = if (food.isFavorite) 1L else 0L,
                        barcode = food.barcode,
                        jsonData = json.encodeToString(food),
                    )
                }
            }
            if (page.size < PAGE_SIZE) break
            offset += PAGE_SIZE
        }
    }

    private suspend fun downloadRecipes() {
        val queries = db.userDataDatabaseQueries
        val summaries = api.getRecipes()
        // Details carry the ingredient list — required for local editing. One
        // request per recipe is unavoidable, so run them in small parallel
        // batches instead of a strictly sequential walk.
        for (batch in summaries.chunked(DETAIL_CONCURRENCY)) {
            val details =
                coroutineScope {
                    batch
                        .map { summary -> async { api.getRecipe(summary.id) } }
                        .awaitAll()
                        .map { it.copy(imageUrl = localizedImageUrl(it.imageUrl)) }
                }
            queries.transaction {
                for (recipe in details) {
                    queries.insertRecipe(
                        id = recipe.id,
                        name = recipe.name,
                        totalServings = recipe.totalServings,
                        isFavorite = if (recipe.isFavorite) 1L else 0L,
                        calories = recipe.calories,
                        protein = recipe.protein,
                        carbs = recipe.carbs,
                        fat = recipe.fat,
                        fiber = recipe.fiber,
                        jsonData = json.encodeToString(recipe),
                    )
                }
            }
        }
    }

    private suspend fun downloadSupplements() {
        val queries = db.userDataDatabaseQueries
        // `all=true`: the default list is active-only, and archived supplements
        // are just as unrecoverable as any other row once the account is gone.
        val supplements = api.getAllSupplements().supplements
        queries.transaction {
            for (supplement in supplements) {
                queries.insertSupplement(
                    id = supplement.id,
                    name = supplement.name,
                    isActive = if (supplement.isActive) 1L else 0L,
                    sortOrder = supplement.sortOrder.toLong(),
                    jsonData = json.encodeToString(supplement),
                )
            }
        }
    }

    /**
     * Entries, supplement logs, sleep and day properties in one walk over the
     * date range — they share the same windows, so four separate walks would
     * only quadruple the round trips.
     */
    private suspend fun downloadDatedCollections(
        floor: String,
        ceiling: String,
    ) {
        val queries = db.userDataDatabaseQueries
        forEachWindow(floor, ceiling) { from, to ->
            val entries = api.getEntriesRange(from, to)
            val supplementLogs = api.getSupplementHistory(from, to).history
            val sleep = api.getSleepEntries(from, to)
            val days = api.getDayPropertiesRange(from, to)
            queries.transaction {
                for (item in entries) {
                    val entry = item.toEntry()
                    queries.insertEntry(
                        id = entry.id,
                        date = entry.date,
                        mealType = entry.mealType,
                        servings = entry.servings,
                        foodId = entry.foodId,
                        recipeId = entry.recipeId,
                        foodName = entry.foodName,
                        calories = entry.calories ?: 0.0,
                        protein = entry.protein ?: 0.0,
                        carbs = entry.carbs ?: 0.0,
                        fat = entry.fat ?: 0.0,
                        fiber = entry.fiber ?: 0.0,
                        jsonData = json.encodeToString(entry),
                    )
                }
                for (log in supplementLogs) {
                    queries.insertSupplementLog(
                        // Same synthesized key as SupplementRepository's cache rows
                        id = "${log.supplementId}-${log.date}",
                        supplementId = log.supplementId,
                        date = log.date,
                        takenAt = log.takenAt,
                    )
                }
                for (entry in sleep) {
                    queries.insertSleepEntry(
                        id = entry.id,
                        entryDate = entry.entryDate,
                        durationMinutes = entry.durationMinutes.toLong(),
                        quality = entry.quality.toLong(),
                        loggedAt = entry.loggedAt,
                        jsonData = json.encodeToString(entry),
                    )
                }
                for (day in days) {
                    queries.upsertDayProperties(
                        date = day.date,
                        isFastingDay = if (day.isFastingDay) 1L else 0L,
                    )
                }
            }
        }
    }

    private suspend fun downloadWeight() {
        val queries = db.userDataDatabaseQueries
        val entries = api.getWeightEntries()
        queries.transaction {
            for (entry in entries) {
                queries.insertWeightEntry(
                    id = entry.id,
                    entryDate = entry.entryDate,
                    weightKg = entry.weightKg,
                    loggedAt = entry.loggedAt,
                    jsonData = json.encodeToString(entry),
                )
            }
        }
    }

    private suspend fun downloadGoals() {
        val goals = api.getGoals() ?: return
        db.userDataDatabaseQueries.insertGoals(
            calorieGoal = goals.calorieGoal,
            proteinGoal = goals.proteinGoal,
            carbGoal = goals.carbGoal,
            fatGoal = goals.fatGoal,
            fiberGoal = goals.fiberGoal,
        )
    }

    private suspend fun downloadPreferences() {
        val preferences = api.getPreferences()
        db.userDataDatabaseQueries.insertPreferences(json.encodeToString(preferences))
    }

    /** Runs [block] over the range in windows the server accepts (≤366 days). */
    private suspend fun forEachWindow(
        floor: String,
        ceiling: String,
        block: suspend (from: String, to: String) -> Unit,
    ) {
        val start = LocalDate.parse(floor)
        var to = maxOf(LocalDate.parse(ceiling), start)
        while (to >= start) {
            val from = maxOf(to.minus(DatePeriod(days = WINDOW_DAYS - 1)), start)
            block(from.toString(), to.toString())
            to = from.minus(DatePeriod(days = 1))
        }
    }

    private fun EntryRangeItem.toEntry(): Entry =
        Entry(
            id = id,
            foodId = foodId,
            recipeId = recipeId,
            supplementId = supplementId,
            date = date,
            mealType = mealType,
            servings = servings,
            notes = notes,
            quickName = quickName,
            quickCalories = quickCalories,
            quickProtein = quickProtein,
            quickCarbs = quickCarbs,
            quickFat = quickFat,
            quickFiber = quickFiber,
            quickNutrients = quickNutrients,
            eatenAt = eatenAt,
            foodName = foodName,
            calories = calories,
            protein = protein,
            carbs = carbs,
            fat = fat,
            fiber = fiber,
            servingSize = servingSize,
            servingUnit = servingUnit,
        )

    companion object {
        private const val PAGE_SIZE = 200
        private const val WINDOW_DAYS = 360
        private const val DETAIL_CONCURRENCY = 6

        /** Only reached when the server reports neither a creation date nor any dated row. */
        private const val FALLBACK_FLOOR = "2024-01-01"
    }
}

File: mobile/iosApp/Bissbilanz/Migration/AccountDowngrader.swift
import Foundation
import SwiftData

/// Turns a synced account into a local-only install: downloads the complete
/// account history into the SwiftData store, then deletes the server account
/// and flips the app to `AppMode.local` — the user keeps all their data
/// on-device. The inverse of `LocalDataMigrator` (which uploads local data on
/// sign-in); because the download writes canonical server rows (real UUIDs, no
/// `temp_` ids), a later sign-in simply migrates them back up through the
/// normal migration path. Mirrors the Android `AccountDowngrader`.
@MainActor
final class AccountDowngrader {
    enum Phase {
        case syncing
        case downloading
        case deleting
    }

    enum DowngradeError: Error {
        case pendingChanges
    }

    private let api: BissbilanzAPI
    private let context: ModelContext
    private let syncManager: SyncManager
    private let authManager: AuthManager
    private let appModeManager: AppModeManager

    private static let windowDays = 360
    /// Rows written between main-thread yields. The store replacement runs on
    /// the main actor (the app's only `ModelContext`), so it has to hand the run
    /// loop back regularly — a multi-year account is tens of thousands of rows,
    /// and one uninterrupted write is a frozen UI and a watchdog kill.
    private static let insertChunk = 200

    init(
        api: BissbilanzAPI,
        context: ModelContext,
        syncManager: SyncManager,
        authManager: AuthManager,
        appModeManager: AppModeManager
    ) {
        self.api = api
        self.context = context
        self.syncManager = syncManager
        self.authManager = authManager
        self.appModeManager = appModeManager
    }

    func downgrade(onPhase: (Phase) -> Void) async throws {
        onPhase(.syncing)
        try await drainPendingQueue()

        onPhase(.downloading)
        let account = try await api.getAccount()
        let today = DateFormatting.today
        let createdOn = account.user.createdAt.map { String($0.prefix(10)) }
        // The account's own dated rows bound the download, NOT its creation
        // date: entry/sleep/weight/day dates are client-chosen, so imported or
        // backfilled days legitimately predate the account, and a device running
        // ahead of UTC can log a day past the server's "today".
        let floor = [account.dataRange.earliest, createdOn].compactMap { $0 }.min() ?? "2024-01-01"
        let ceiling = max(account.dataRange.latest ?? today, today)

        // Fetch the complete account into memory first; the store is only
        // touched once everything arrived, so a failed download changes nothing.
        var foods: [Food] = []
        var offset = 0
        while true {
            let page = try await api.getFoods(limit: 200, offset: offset)
            foods.append(contentsOf: page)
            if page.count < 200 { break }
            offset += 200
        }

        var recipes: [Recipe] = []
        for summary in try await api.getRecipes() {
            // Details carry the ingredient list — required for local editing.
            recipes.append(try await api.getRecipe(id: summary.id))
        }

        // `all: true` — archived supplements are as unrecoverable as any other
        // row once the account is deleted.
        let supplements = try await api.getSupplements(all: true)

        var entries: [Entry] = []
        var supplementLogs: [SupplementHistoryEntry] = []
        var sleepEntries: [SleepEntry] = []
        var dayProperties: [DayProperties] = []
        try await forEachWindow(floor: floor, ceiling: ceiling) { from, to in
            entries.append(contentsOf: try await self.api.getEntriesRange(startDate: from, endDate: to))
            supplementLogs.append(
                contentsOf: try await self.api.getSupplementHistory(startDate: from, endDate: to)
            )
            sleepEntries.append(contentsOf: try await self.api.getSleepEntries(from: from, to: to))
            dayProperties.append(
                contentsOf: try await self.api.getDayPropertiesRange(startDate: from, endDate: to)
            )
        }

        // Server-hosted photos die with the account, so pull each one into the
        // App Group image store and repoint the row at the local file. A photo
        // that can't be fetched loses its reference rather than keeping a URL
        // that will 404 forever.
        let localizedFoods = await localizePhotos(foods, url: \.imageUrl) { food, url in
            try? JSONPatch.merged(Food.self, base: food, patch: ["imageUrl": url as Any])
        }
        let localizedRecipes = await localizePhotos(recipes, url: \.imageUrl) { recipe, url in
            try? JSONPatch.merged(Recipe.self, base: recipe, patch: ["imageUrl": url as Any])
        }

        let weightEntries = try await api.getWeightEntries()
        let goals = try await api.getGoals()
        let preferences = try await api.getPreferences()

        // Replace the cache wholesale — the sync queue is empty (checked
        // above), so there are no optimistic local rows to preserve. The
        // batch deletes below are applied straight to the persistent store and
        // are NOT undone by a failing `save()`, so anything that throws from
        // here on leaves a partially emptied cache. That is recoverable — the
        // account still exists (it is deleted further down) and the next
        // refresh re-fills the store — but it must not be silent, so the
        // deletes propagate their errors instead of being swallowed.
        try context.delete(model: LocalEntry.self)
        try context.delete(model: LocalFood.self)
        try context.delete(model: LocalRecipe.self)
        try context.delete(model: LocalWeightEntry.self)
        try context.delete(model: LocalSleepEntry.self)
        try context.delete(model: LocalSupplement.self)
        try context.delete(model: LocalSupplementLog.self)
        try context.delete(model: LocalGoals.self)
        try context.delete(model: LocalPreferences.self)
        try context.delete(model: LocalDayProperties.self)

        await insert(localizedFoods) { LocalFood(food: $0) }
        await insert(localizedRecipes) { LocalRecipe(recipe: $0) }
        await insert(supplements) { LocalSupplement(supplement: $0) }
        await insert(entries) { LocalEntry(entry: $0, date: $0.date ?? today) }
        await insert(supplementLogs) {
            LocalSupplementLog(supplementId: $0.supplementId, date: $0.date, takenAt: $0.takenAt)
        }
        await insert(weightEntries) { LocalWeightEntry(entry: $0) }
        await insert(sleepEntries) { LocalSleepEntry(entry: $0) }
        await insert(dayProperties) { LocalDayProperties(properties: $0) }
        if let goals {
            context.insert(LocalGoals(goals: goals))
        }
        context.insert(LocalPreferences(preferences: preferences))
        try context.save()

        onPhase(.deleting)
        try await finalizeDowngrade()
    }

    /// Empties the sync queue, or throws. A single drain uploads only the
    /// operations that are due, so a device that was offline for a while needs
    /// several passes — but an operation that never uploads must not spin here
    /// forever, hence "stop as soon as a pass makes no progress".
    private func drainPendingQueue() async throws {
        var pending = syncManager.pendingCount
        while pending > 0 {
            _ = await syncManager.drainPendingQueue()
            let remaining = syncManager.pendingCount
            if remaining >= pending {
                throw DowngradeError.pendingChanges
            }
            pending = remaining
        }
    }

    /// Deletes the server account and switches the app to Local mode.
    ///
    /// Step order is load-bearing and must not be "improved": authenticated +
    /// `.local` is what routes the app to the migration screen, so the mode may
    /// only flip once the session is gone. Everything after the delete is
    /// therefore ordered so it cannot leave the app in `.synced` against an
    /// account that no longer exists — that state is how the user ends up
    /// signing out and wiping the data this download just rescued.
    private func finalizeDowngrade() async throws {
        try await api.deleteAccount()
        syncManager.clearQueue()
        authManager.logout()
        appModeManager.setMode(.local)
    }

    /// Downloads every `/uploads/` photo in `items` into `LocalImageStore` and
    /// rewrites the row's URL to the resulting `file://` path. Rows without a
    /// server-hosted photo (none, or a public Open Food Facts URL) pass through
    /// untouched, as do rows whose download fails.
    private func localizePhotos<T>(
        _ items: [T],
        url: KeyPath<T, String?>,
        rewrite: (T, String) -> T?
    ) async -> [T] {
        var result: [T] = []
        result.reserveCapacity(items.count)
        for item in items {
            guard let imageUrl = item[keyPath: url],
                  let key = LocalImageStore.cacheKey(for: imageUrl)
            else {
                result.append(item)
                continue
            }
            var file = LocalImageStore.cachedFile(for: imageUrl)
            if file == nil, let data = try? await api.downloadImage(path: imageUrl) {
                file = LocalImageStore.write(data, named: key)
            }
            guard let file else {
                result.append(item)
                continue
            }
            result.append(rewrite(item, file.absoluteString) ?? item)
        }
        return result
    }

    private func insert<T, M: PersistentModel>(_ items: [T], _ make: (T) -> M) async {
        for (index, item) in items.enumerated() {
            context.insert(make(item))
            if index % Self.insertChunk == Self.insertChunk - 1 {
                await Task.yield()
            }
        }
    }

    /// Runs `block` over the range in windows the server accepts (≤366 days).
    private func forEachWindow(
        floor: String,
        ceiling: String,
        block: (String, String) async throws -> Void
    ) async throws {
        guard let start = DateFormatting.date(from: floor),
              let end = DateFormatting.date(from: ceiling)
        else { return }
        var to = max(end, start)
        let calendar = Calendar.current
        while to >= start {
            let windowStart = calendar.date(byAdding: .day, value: -(Self.windowDays - 1), to: to) ?? start
            let from = max(windowStart, start)
            try await block(DateFormatting.isoString(from: from), DateFormatting.isoString(from: to))
            guard let next = calendar.date(byAdding: .day, value: -1, to: from) else { break }
            to = next
        }
    }
}
